### ISSUE TITLE
Clarify impact of warnings from Vc package

### ISSUE THREAD

**Issue by matthiasrichter** [AliceO2Group/AliceO2#340]
I see the following warnings from the compilation of the Vc package, we need to clarify the impact to AliceO2. If there is no impact, the Vc headers should be included in the `SYSTEMINCLUDE_DIRECTORIES` section to mask the warnings

```
In file included from input_line_12:9:
In file included from /src/src/alisw/sw/ubuntu1604_x86-64/Vc/1.2.0_O2-3/include/Vc/Vc:31:
In file included from /src/src/alisw/sw/ubuntu1604_x86-64/Vc/1.2.0_O2-3/include/Vc/vector.h:38:
In file included from /src/src/alisw/sw/ubuntu1604_x86-64/Vc/1.2.0_O2-3/include/Vc/scalar/../common/../sse/vector.h:33:
In file included from /src/src/alisw/sw/ubuntu1604_x86-64/Vc/1.2.0_O2-3/include/Vc/scalar/../common/../sse/intrinsics.h:54:
/src/src/alisw/sw/ubuntu1604_x86-64/Vc/1.2.0_O2-3/include/Vc/scalar/../common/../sse/const_data.h:70:103: warning: 
      instantiation of variable 'Vc_1::SSE::c_log<float>::data' required here, but no
      definition is available [-Wundefined-var-template]
  ...const float *d(int i) { return reinterpret_cast<const  float *>(&data[i * Size...
                                                                      ^
/src/src/alisw/sw/ubuntu1604_x86-64/Vc/1.2.0_O2-3/include/Vc/scalar/../common/../sse/vectorhelper.h:261:126: note: 
      in instantiation of member function 'Vc_1::SSE::c_log<float>::d' requested here
  ..._mm_castps_si128(_mm_load_ps(c_log<float>::d(1)))));
                                                ^
/src/src/alisw/sw/ubuntu1604_x86-64/Vc/1.2.0_O2-3/include/Vc/scalar/../common/../sse/const_data.h:71:43: note: 
      forward declaration of template entity is here
    alignas(64) static const unsigned int data[];
                                          ^
/src/src/alisw/sw/ubuntu1604_x86-64/Vc/1.2.0_O2-3/include/Vc/scalar/../common/../sse/const_data.h:70:103: note: 
      add an explicit instantiation declaration to suppress this warning if
      'Vc_1::SSE::c_log<float>::data' is explicitly instantiated in another
      translation unit
  ...const float *d(int i) { return reinterpret_cast<const  float *>(&data[i * Size...
                                                                      ^
```

**Comment by matthiasrichter** [comment]
@mattkretz could you have a look what the warnings are about? Is this some issue of the AliceO2 code? Thanks a lot. Matthias

**Comment by mattkretz** [comment]
I've seen the warning before. This should be the fix: https://github.com/VcDevel/Vc/commit/22845392a5ac7aea59ebecd23075b1b93a7b6eae
It's part of the latest Vc release.

**Comment by matthiasrichter** [comment]
thanks a lot, Matthias!

**Comment by matthiasrichter** [comment]
It turns out that the warnings are thrown in the dictionary generation of the TPCSimulation and defining the VC includes as system headers does not work somehow to suppress the warning.

On the other hand we can not just switch to a more recent version containing the fix because this has implications to the build of AliRoot as well and the deployment of software for analysis.

**Comment by mattkretz** [comment]
> On the other hand we can not just switch to a more recent version containing the fix because this has implications to the build of AliRoot as well and the deployment of software for analysis.

IIUC you're saying that fixing bugs (or warnings in this case) in dependent libraries is not possible? That sounds like a recipe for disaster. You've got to have a process for updating dependencies...

**Comment by matthiasrichter** [comment]
I did not complete my message, we will not switch immediately. But very soon, there is just more thinks and implications to think about. The issue stays open until it's fixed. Sure.

**Comment by dberzano** [comment]
Hi all, alisw/alidist#695 (just merged) introduces C++14 support and updates Vc to 1.3.0 (not 1.3.1 though, we can do this minor bump too).

In particular, Vc will be updated for both Run2 and Run3 software, as AFAIK GCC 6 cannot compile Vc 1.2 and our only users (HLT, @mkrzewic) green-lighted the update.

Can you test with the new master from `alidist` and see if the warnings are still there?

**Comment by mkrzewic** [comment]
Hi @dberzano in the production system we use Vc 1.3.1, and since the move to 1.3 I have not seen the warnings, but your mileage might vary, as always...

**Comment by matthiasrichter** [comment]
Hi, the warning is only fixed in tag 1.3.1

**Comment by matthiasrichter** [comment]
unfortunately, the situation got worse by orders of magnitude in the new recipe, whether it's due to Vc 1.3.0 or gcc 6.2 or both needs to be clarified. There are tons of warnings like that:
```bash
In file included from /src/src/alisw/sw/ubuntu1604_x86-64/Vc/1.3.0_O2-1/include/Vc/vector.h:191:0,
                 from /src/src/alisw/sw/ubuntu1604_x86-64/Vc/1.3.0_O2-1/include/Vc/Vc:30,
                 from /home/richter/src/alisw/sw/SOURCES/O2/dev/0/Detectors/TPC/base/include/TPCBase/CalArray.h:4,
                 from /home/richter/src/alisw/sw/BUILD/c44aa19ace8fc3304be82ad541921a0969c06d00/O2/Detectors/TPC/base/G__TPCBaseDict.cxx:40:
/src/src/alisw/sw/ubuntu1604_x86-64/Vc/1.3.0_O2-1/include/Vc/common/simdarray.h: In instantiation of ‘class Vc_1::SimdArray<int, 2ul, Vc_1::Vector<int, Vc_1::VectorAbi::Scalar>, 1ul>’:
/src/src/alisw/sw/ubuntu1604_x86-64/Vc/1.3.0_O2-1/include/Vc/sse/math.h:70:11:   required from here
/src/src/alisw/sw/ubuntu1604_x86-64/Vc/1.3.0_O2-1/include/Vc/common/simdarray.h:569:81: warning: ignoring attributes on template argument ‘Vc_1::Vector<int, Vc_1::VectorAbi::Scalar>::EntryType {aka int}’ [-Wignored-attributes]
         std::is_same<typename V::EntryType, typename V::VectorEntryType>::value ||
         ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~^~
             // or N is a multiple of V::size()
             ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~                                   
             (N % V::size() == 0),
             ~~~~~~~~~~~~~~~~~~~~                                                 
/src/src/alisw/sw/ubuntu1604_x86-64/Vc/1.3.0_O2-1/include/Vc/common/simdarray.h:569:81: warning: ignoring attributes on template argument ‘Vc_1::Vector<int, Vc_1::VectorAbi::Scalar>::VectorEntryType {aka int}’ [-Wignored-attributes]

/src/src/alisw/sw/ubuntu1604_x86-64/Vc/1.3.0_O2-1/include/Vc/common/operators.h: In substitution of ‘template<class V, class T> using ReturnType = Vc_1::Detail::ReturnTypeImpl<V, T, (std::is_arithmetic<U>::value || std::is_convertible<T, int>::value), decltype (is_convertible_to_any_vector<typename V::value_type, typename V::abi>(declval<const T&>())), std::is_integral<typename V::value_type>::value> [with V = Vc_1::Vector<double, Vc_1::VectorAbi::Sse>; T = Vc_1::Vector<double, Vc_1::VectorAbi::Sse>]’:
/src/src/alisw/sw/ubuntu1604_x86-64/Vc/1.3.0_O2-1/include/Vc/common/operators.h:249:1:   required from ‘Vc_1::enable_if<(std::is_convertible<Vc_1::Vector<T, Abi>, typename Vc_1::Detail::ReturnTypeImpl<Vc_1::Vector<T, Abi>, U, (std::is_arithmetic<U>::value || std::is_convertible<U, int>::value), decltype (is_convertible_to_any_vector<typename Vc_1::Vector<T, Abi>::value_type, typename Vc_1::Vector<T, Abi>::abi>(declval<const U&>())), std::is_integral<typename Vc_1::Vector<T, Abi>::value_type>::value>::type>::value && std::is_convertible<U, typename Vc_1::Detail::ReturnTypeImpl<Vc_1::Vector<T, Abi>, U, (std::is_arithmetic<U>::value || std::is_convertible<U, int>::value), decltype (is_convertible_to_any_vector<typename Vc_1::Vector<T, Abi>::value_type, typename Vc_1::Vector<T, Abi>::abi>(declval<const U&>())), std::is_integral<typename Vc_1::Vector<T, Abi>::value_type>::value>::type>::value), typename Vc_1::Detail::ReturnTypeImpl<Vc_1::Vector<T, Abi>, U, (std::is_arithmetic<U>::value || std::is_convertible<U, int>::value), decltype (is_convertible_to_any_vector<typename Vc_1::Vector<T, Abi>::value_type, typename Vc_1::Vector<T, Abi>::abi>(declval<const U&>())), std::is_integral<typename Vc_1::Vector<T, Abi>::value_type>::value>::type::Mask> Vc_1::operator==(Vc_1::Vector<T, Abi>, const U&) [with T = double; Abi = Vc_1::VectorAbi::Sse; U = Vc_1::Vector<double, Vc_1::VectorAbi::Sse>; Vc_1::enable_if<(std::is_convertible<Vc_1::Vector<T, Abi>, typename Vc_1::Detail::ReturnTypeImpl<Vc_1::Vector<T, Abi>, U, (std::is_arithmetic<U>::value || std::is_convertible<U, int>::value), decltype (is_convertible_to_any_vector<typename Vc_1::Vector<T, Abi>::value_type, typename Vc_1::Vector<T, Abi>::abi>(declval<const U&>())), std::is_integral<typename Vc_1::Vector<T, Abi>::value_type>::value>::type>::value && std::is_convertible<U, typename Vc_1::Detail::ReturnTypeImpl<Vc_1::Vector<T, Abi>, U, (std::is_arithmetic<U>::value || std::is_convertible<U, int>::value), decltype (is_convertible_to_any_vector<typename Vc_1::Vector<T, Abi>::value_type, typename Vc_1::Vector<T, Abi>::abi>(declval<const U&>())), std::is_integral<typename Vc_1::Vector<T, Abi>::value_type>::value>::type>::value), typename Vc_1::Detail::ReturnTypeImpl<Vc_1::Vector<T, Abi>, U, (std::is_arithmetic<U>::value || std::is_convertible<U, int>::value), decltype (is_convertible_to_any_vector<typename Vc_1::Vector<T, Abi>::value_type, typename Vc_1::Vector<T, Abi>::abi>(declval<const U&>())), std::is_integral<typename Vc_1::Vector<T, Abi>::value_type>::value>::type::Mask> = Vc_1::Mask<double, Vc_1::VectorAbi::Sse>]’
/src/src/alisw/sw/ubuntu1604_x86-64/Vc/1.3.0_O2-1/include/Vc/sse/math.h:67:55:   required from here
/src/src/alisw/sw/ubuntu1604_x86-64/Vc/1.3.0_O2-1/include/Vc/common/operators.h:125:53: warning: ignoring attributes on template argument ‘Vc_1::Vector<double, Vc_1::VectorAbi::Sse>::value_type {aka double}’ [-Wignored-attributes]
     std::is_integral<typename V::value_type>::value>;
                                                     ^
```

**Comment by sawenzel** [comment]
Absolutely agree. Was about to write the same.

**Comment by mattkretz** [comment]
the warning is new since GCC6 and produces tons of false positives. Only sensible way is to disable it. The flags produced by VcConfig should have that flag. But maybe I forgot. I'll have to take a look next week

On April 28, 2017 11:00:15 PM GMT+02:00, Matthias Richter <notifications@github.com> wrote:
>unfortunately, the situation got worse by orders of magnitude in the
>new recipe, whether it's due to Vc 1.3.0 or gcc 6.2 or both needs to be
>clarified. There are tons of warnings like that:
>```bash
>In file included from
>/src/src/alisw/sw/ubuntu1604_x86-64/Vc/1.3.0_O2-1/include/Vc/vector.h:191:0,
>from
>/src/src/alisw/sw/ubuntu1604_x86-64/Vc/1.3.0_O2-1/include/Vc/Vc:30,
>from
>/home/richter/src/alisw/sw/SOURCES/O2/dev/0/Detectors/TPC/base/include/TPCBase/CalArray.h:4,
>from
>/home/richter/src/alisw/sw/BUILD/c44aa19ace8fc3304be82ad541921a0969c06d00/O2/Detectors/TPC/base/G__TPCBaseDict.cxx:40:
>/src/src/alisw/sw/ubuntu1604_x86-64/Vc/1.3.0_O2-1/include/Vc/common/simdarray.h:
>In instantiation of ‘class Vc_1::SimdArray<int, 2ul, Vc_1::Vector<int,
>Vc_1::VectorAbi::Scalar>, 1ul>’:
>/src/src/alisw/sw/ubuntu1604_x86-64/Vc/1.3.0_O2-1/include/Vc/sse/math.h:70:11:
>  required from here
>/src/src/alisw/sw/ubuntu1604_x86-64/Vc/1.3.0_O2-1/include/Vc/common/simdarray.h:569:81:
>warning: ignoring attributes on template argument ‘Vc_1::Vector<int,
>Vc_1::VectorAbi::Scalar>::EntryType {aka int}’ [-Wignored-attributes]
>std::is_same<typename V::EntryType, typename V::VectorEntryType>::value
>||
>~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~^~
>             // or N is a multiple of V::size()
>  ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~                                   
>             (N % V::size() == 0),
>  ~~~~~~~~~~~~~~~~~~~~                                                 
>/src/src/alisw/sw/ubuntu1604_x86-64/Vc/1.3.0_O2-1/include/Vc/common/simdarray.h:569:81:
>warning: ignoring attributes on template argument ‘Vc_1::Vector<int,
>Vc_1::VectorAbi::Scalar>::VectorEntryType {aka int}’
>[-Wignored-attributes]
>
>/src/src/alisw/sw/ubuntu1604_x86-64/Vc/1.3.0_O2-1/include/Vc/common/operators.h:
>In substitution of ‘template<class V, class T> using ReturnType =
>Vc_1::Detail::ReturnTypeImpl<V, T, (std::is_arithmetic<U>::value ||
>std::is_convertible<T, int>::value), decltype
>(is_convertible_to_any_vector<typename V::value_type, typename
>V::abi>(declval<const T&>())), std::is_integral<typename
>V::value_type>::value> [with V = Vc_1::Vector<double,
>Vc_1::VectorAbi::Sse>; T = Vc_1::Vector<double,
>Vc_1::VectorAbi::Sse>]’:
>/src/src/alisw/sw/ubuntu1604_x86-64/Vc/1.3.0_O2-1/include/Vc/common/operators.h:249:1:
>required from ‘Vc_1::enable_if<(std::is_convertible<Vc_1::Vector<T,
>Abi>, typename Vc_1::Detail::ReturnTypeImpl<Vc_1::Vector<T, Abi>, U,
>(std::is_arithmetic<U>::value || std::is_convertible<U, int>::value),
>decltype (is_convertible_to_any_vector<typename Vc_1::Vector<T,
>Abi>::value_type, typename Vc_1::Vector<T, Abi>::abi>(declval<const
>U&>())), std::is_integral<typename Vc_1::Vector<T,
>Abi>::value_type>::value>::type>::value && std::is_convertible<U,
>typename Vc_1::Detail::ReturnTypeImpl<Vc_1::Vector<T, Abi>, U,
>(std::is_arithmetic<U>::value || std::is_convertible<U, int>::value),
>decltype (is_convertible_to_any_vector<typename Vc_1::Vector<T,
>Abi>::value_type, typename Vc_1::Vector<T, Abi>::abi>(declval<const
>U&>())), std::is_integral<typename Vc_1::Vector<T,
>Abi>::value_type>::value>::type>::value), typename
>Vc_1::Detail::ReturnTypeImpl<Vc_1::Vector<T, Abi>, U,
>(std::is_arithmetic<U>::value || std::is_convertible<U, int>::value),
>decltype (is_convertible_to_any_vector<typename Vc_1::Vector<T,
>Abi>::value_type, typename Vc_1::Vector<T, Abi>::abi>(declval<const
>U&>())), std::is_integral<typename Vc_1::Vector<T,
>Abi>::value_type>::value>::type::Mask> Vc_1::operator==(Vc_1::Vector<T,
>Abi>, const U&) [with T = double; Abi = Vc_1::VectorAbi::Sse; U =
>Vc_1::Vector<double, Vc_1::VectorAbi::Sse>;
>Vc_1::enable_if<(std::is_convertible<Vc_1::Vector<T, Abi>, typename
>Vc_1::Detail::ReturnTypeImpl<Vc_1::Vector<T, Abi>, U,
>(std::is_arithmetic<U>::value || std::is_convertible<U, int>::value),
>decltype (is_convertible_to_any_vector<typename Vc_1::Vector<T,
>Abi>::value_type, typename Vc_1::Vector<T, Abi>::abi>(declval<const
>U&>())), std::is_integral<typename Vc_1::Vector<T,
>Abi>::value_type>::value>::type>::value && std::is_convertible<U,
>typename Vc_1::Detail::ReturnTypeImpl<Vc_1::Vector<T, Abi>, U,
>(std::is_arithmetic<U>::value || std::is_convertible<U, int>::value),
>decltype (is_convertible_to_any_vector<typename Vc_1::Vector<T,
>Abi>::value_type, typename Vc_1::Vector<T, Abi>::abi>(declval<const
>U&>())), std::is_integral<typename Vc_1::Vector<T,
>Abi>::value_type>::value>::type>::value), typename
>Vc_1::Detail::ReturnTypeImpl<Vc_1::Vector<T, Abi>, U,
>(std::is_arithmetic<U>::value || std::is_convertible<U, int>::value),
>decltype (is_convertible_to_any_vector<typename Vc_1::Vector<T,
>Abi>::value_type, typename Vc_1::Vector<T, Abi>::abi>(declval<const
>U&>())), std::is_integral<typename Vc_1::Vector<T,
>Abi>::value_type>::value>::type::Mask> = Vc_1::Mask<double,
>Vc_1::VectorAbi::Sse>]’
>/src/src/alisw/sw/ubuntu1604_x86-64/Vc/1.3.0_O2-1/include/Vc/sse/math.h:67:55:
>  required from here
>/src/src/alisw/sw/ubuntu1604_x86-64/Vc/1.3.0_O2-1/include/Vc/common/operators.h:125:53:
>warning: ignoring attributes on template argument ‘Vc_1::Vector<double,
>Vc_1::VectorAbi::Sse>::value_type {aka double}’ [-Wignored-attributes]
>     std::is_integral<typename V::value_type>::value>;
>                                                     ^
>```
>
>-- 
>You are receiving this because you were mentioned.
>Reply to this email directly or view it on GitHub:
>https://github.com/AliceO2Group/AliceO2/issues/340#issuecomment-298105468

-- 
Dr. Matthias Kretz
Senior Software Engineer
Experiment Systems

phone: +49 6159 713084
e-mail: m.kretz@gsi.de

GSI Helmholtzzentrum für Schwerionenforschung GmbH 
Planckstraße 1 
64291 Darmstadt 
www.gsi.de 

Gesellschaft mit beschränkter Haftung 
Sitz der Gesellschaft: Darmstadt 
Handelsregister: Amtsgericht Darmstadt, HRB 1528 

Geschäftsführung: 
Ursula Weyrich 
Professor Dr. Karlheinz Langanke 

Vorsitzender des Aufsichtsrates: St Dr. Georg Schütte 
Stellvertreter: Ministerialdirigent Dr. Rolf Bernhardt

**Comment by mattkretz** [comment]
OK, so the flag is really only used for compiling Vc itself. So, to have a proper fix, we'll need a Vc 1.3.2 release anyway. I opened VcDevel/Vc#164.

**Comment by mattkretz** [comment]
could you test latest Vc master whether all the warnings are resolved? If not I can't guarantee that 1.3.2 will be silent ;-)

**Comment by matthiasrichter** [comment]
great! will test it now. Thanks.

**Comment by matthiasrichter** [comment]
Hi Matthias, all the warnings are gone, thanks. So whenever a new tag is available, leave a note here. Then the recipe can be updated.

**Comment by mattkretz** [comment]
perfect. I'll make a release then

**Comment by dberzano** [comment]
Fantastic, we'll upload the alidist recipe accordingly when done! Thanks @mattkretz!

**Comment by mattkretz** [comment]
here you go: https://github.com/VcDevel/Vc/releases/tag/1.3.2

**Comment by dberzano** [comment]
Alright, if you update your `alidist` recipes you'll get Vc 1.3.2 for all O2 defaults.

**Comment by matthiasrichter** [comment]
I'm closing the issue, everything solved in Vc 1.3.2, thanks for the quick response.
